scripts/Diffraction/isis_powder/routines/InstrumentSettings.py

```diff
@@ -0,0 +1,108 @@
+        self._basic_conf_dict = basic_conf_dict
+        self._kwargs = kwargs
+
+        self._unknown_keys_found = False
+        self._parse_attributes(dict_to_parse=adv_conf_dict)
+        self._parse_attributes(dict_to_parse=basic_conf_dict)
+        self._parse_attributes(dict_to_parse=kwargs)
+        if self._unknown_keys_found:
+            _print_known_keys(attr_mapping)
+
+    def __getattr__(self, item):
+        map_entry = next((attr_tuple for attr_tuple in self._attr_mapping if item == attr_tuple[-1]), None)
+        if map_entry:
+            # User forgot to enter the param:
+            raise AttributeError("The parameter with name: '" + str(map_entry[0]) + "' is required but was not set or "
+                                 "passed.\nPlease set this configuration option and try again")
+        else:
+            # If you have got here from a grep or something similar this error message means the line caller
+            # has asked for a class attribute which does not exist. These attributes are set in a mapping file which
+            # is passed in whilst InstrumentSettings is being constructed. Check that the 'script name' (i.e. not user
+            # friendly name) is typed correctly in both the script(s) and mapping file.
+            raise AttributeError("The attribute in the script with name " + str(item) + " is unknown to the mapping."
+                                 "\nPlease contact the development team.")
+
+    def check_expected_attributes_are_set(self, expected_attr_names):
+        for expected_attr in expected_attr_names:
+            if not [attr_entry for attr_entry in self._attr_mapping if expected_attr == attr_entry[-1]]:
+                raise ValueError("Expected attribute '" + str(expected_attr) + "' is unknown to attribute mapping")
+
+        # Filter down the full mapping list
+        found_tuple_list = [tuple_entry for tuple_entry in self._attr_mapping if tuple_entry[-1] in expected_attr_names]
+        expected_params_dict = dict(found_tuple_list)
+        self._check_attribute_is_set(expected_params_dict)
+
+    def update_attributes(self, advanced_config=None, basic_config=None, kwargs=None, suppress_warnings=False):
+        self._adv_config_dict = advanced_config if advanced_config else self._adv_config_dict
+        self._basic_conf_dict = basic_config if basic_config else self._basic_conf_dict
+        self._kwargs = kwargs if kwargs else self._kwargs
+
+        has_known_keys_already_been_printed = self._unknown_keys_found
+        # Only update if one in hierarchy below it has been updated
+        if advanced_config:
+            self._parse_attributes(self._adv_config_dict, suppress_warnings=suppress_warnings)
+        if advanced_config or basic_config:
+            self._parse_attributes(self._basic_conf_dict,
+                                   suppress_warnings=(not bool(basic_config or suppress_warnings)))
+        if advanced_config or basic_config or kwargs:
```

Comment:
Added comment explaining why its done like this.
Tl;Dr - Its to ensure their overrides are preserved when reloading a config (e.g. executing focus twice)